### ISSUE TITLE
Loose comparison for array_search and in_array?

### ISSUE THREAD

**Issue by dingo-d** [WordPress-Coding-Standards/WordPress-Coding-Standards#829]
Both `array_search()` and `in_array()`, when `true` as third argument is not supplied, will throw an

> Not using strict comparison for _function_name()_; supply true for third argument

warning.

I tried adding 

~~~php
// WPCS: loose comparison ok.
~~~

Afterwards, but the warning still remains.

Should the loose comparison whitelist be applied to those functions as well as `==` operator?

**Comment by jrfnl** [comment]
Out of curiosity: Could you give a code example for something where you would consider a loose comparison for one of these array functions ok ?

**Comment by dingo-d** [comment]
This is for the internal importer I'm working on at my firm, but it's just one example that I can think of atm.

I have 2 arrays both containing menu item ID's. The first one contains relations between old ID's for menu parents in the form of `[old id]=> new_id` for each menu item

~~~php
Array
(
    [12769] => 0
    [12770] => 0
    [12771] => 47
    [12772] => 47
    [12773] => 47
    [12774] => 47
    [12775] => 91
)
~~~

The second one is an array of new menu ID's for each menu item

~~~php
Array
(
    [0] => 0
    [1] => 0
    [2] => 47
    [3] => 47
    [4] => 47
    [5] => 47
    [6] => 91
    [7] => 91
)
~~~

Now, I needed to compare to see if the value of the new menu parent ID is located in the relation array (so that the parent - child menu item order will work in the imported menu). The thing was happening, was that the second array had values that were integers, while the first array had values that were strings.

By using 

~~~php
foreach ( $parent_id as $parent_id_key => $parent_id_value ) {
    $item_id = array_search( $parent_id_value, $parent_relation_array );
}
~~~

I'll get `true` as a result, while using

~~~php
foreach ( $parent_id as $parent_id_key => $parent_id_value ) {
    $item_id = array_search( $parent_id_value, $parent_relation_array, true );
}
~~~

I get all falses.

It could be said that I probably needed to be a bit more careful when combining the arrays when creating the first array (type casting thing), but that's not the point atm.

**Comment by jrfnl** [comment]
> It could be said that I probably needed to be a bit more careful when combining the arrays when creating the first array (type casting thing), but that's not the point atm.

Ok, I can see how a whitelist comment could help you in this case. If others agree, I'd happily add that ability to the sniff.

All the same, you are opening yourself up to bugs as if - for whatever reason -, *other* strings would enter the array, you will run into problems straight away:

~~~php
Array
(
    [12769] => 0
    [12770] => 0
    [12771] => 47
    [12772] => 47
    [12773] => 47
    [12774] => 47
    [12775] => 91
)
~~~

The second one is an array of new menu ID's for each menu item

Let's say your first array remains the same, but the second for some reason would suddenly return

~~~php
Array
(
    [0] => 'my-menu'
    [1] => 'other-menu'
)
~~~

Both of these would cast to int `0` and match.

**Comment by dingo-d** [comment]
Both of those are using `wp_get_nav_menu_items( $menu_id )`, and I'm pulling the ID's from the `ID` and `menu_item_parent` properties of the menu item objects. Those should be numeric values, no?

**Comment by jrfnl** [comment]
> Those should be numeric values, no?

Numeric yes, integers: not necessarily, as that depends on how the calling functions handle the database call. Everything you receive back from a db query is a string unless it's cast to something else after retrieval.

**Comment by dingo-d** [comment]
I mean, I think I should be fine, I have one extra conditional check before hand to see if I'm fetching the right thing.

Basically, when creating arrays I should just put `(int)` in front of the values to be sure I'll have integers in my arrays?

**Comment by dingo-d** [comment]
I don't know, I've been looking through both [wp_get_nav_menu_items](https://developer.wordpress.org/reference/functions/wp_get_nav_menu_items/) and [wp_setup_nav_menu_item](https://developer.wordpress.org/reference/functions/wp_setup_nav_menu_item/) functions, and I don't think that they ever return a non numeric value.
If I'm wrong please correct me (I hope I'm not since I finally got this to work xD)

**Comment by jrfnl** [comment]
> If I'm wrong please correct me (I hope I'm not since I finally got this to work xD)

That presume that you can trust everything you get back from the database, but can you ? From a security perspective, any data from an external source, be it user input, database, json is suspect and should be treated as such.

**Comment by dingo-d** [comment]
Well since I am populating the stuff in the db (with the export I made), I'm rather sure I have it under control. I mean I've worked on this last year for quite some time, and I've had to revisit it in the last two months, but it's constantly checked and tested. So far I haven't found any issues.

I totally understand what you're saying, and I wholeheartedly agree, but in my situation I don't think that something like that could happen.

I am going to pull the array through the callback you've suggested though, just to be safe :D

Thanks for your input, really means a lot to get feedback from more experienced programmer :)
